packages/classes/src/PayloadArray.js

```diff
@@ -0,0 +1,98 @@
+ */
+class PayloadArray {
+	payloads = []
+
+	constructor(...payloads) {
+		[...payloads].forEach((payload) => this.insert(payload))
+	}
+
+	get length() { return this.payloads.length }
+
+	/**
+	 * Insert a Payload into the PayloadArray while maintaining position order.
+	 *
+	 * @param  {Payload} payload The Payload to be inserted.
+	 * @return {PayloadArray}    This PayloadArray to allow for chaining.
+	 */
+	insert = (payload) => {
+		if (!(payload instanceof Payload)) {
+			throw new Error('Attempt to insert a non-Payload to PayloadArray')
+		}
+		const index = this.indexOfPosition(payload.position)
+		this.payloads.splice(index, 0, payload)
+		return this
+	}
+
+	/**
+	 * Find the index of the first Payload whose position is greater than or equal to this one.
+	 *
+	 * NOTE: if no qualifying payloads exist, this will return the length of the PayloadArray
+	 *
+	 * @param  {Number} position The position being searched for.
+	 * @return {Number}          The index of the item, or the PayloadArray length.
+	 */
+	indexOfPosition = (position) => {
+		const recursiveSearch = (sortedPayloads, target, left, right) => {
+			if (left === right) return left
+			const index = Math.floor((left + right) / 2)
+			if (index === sortedPayloads.length) return index
+			const payload = sortedPayloads[index]
+			if (payload.position === target) return index
+			if (payload.position < target) {
+				return recursiveSearch(sortedPayloads, target, index + 1, right)
+			}
+			if (payload.position > target) return recursiveSearch(sortedPayloads, target, left, index)
+			return index
+		}
+		return recursiveSearch(this.payloads, position, 0, this.payloads.length)
+	}
+
+	/**
+	 * Filter out the Payloads of the specified type.
+	 *
+	 * NOTE: This returns a new PayloadArray and does not modify the original.
+	 *
+	 * @param  {String} type  The type of payload to include in the new PayloadArray.
+	 * @return {PayloadArray} A filtered PayloadArray.
+	 */
+	filterByType = (type) => {
+		const filteredPayloadArray = new PayloadArray()
+		filteredPayloadArray.payloads = this.payloads.filter(
+			(payload) => RegExp(`^${type.replace(/\.ANY$/, '.*')}$`).test(payload.type),
+		)
+		return filteredPayloadArray
+	}
+
+	/**
+	 * Filter out the Payloads that exist within a specified duration.
+	 *
+	 * NOTE: This returns a new PayloadArray and does not modify the original.
+	 *
+	 * @param  {Number} start The start position.
+	 * @param  {Number} end   The end position.
+	 * @return {PayloadArray} A filtered PayloadArray.
+	 */
+	filterByPosition = (start, end = null) => {
+		const filteredPayloadArray = new PayloadArray()
+		filteredPayloadArray.payloads = this.payloads.filter(
+			(payload) => (payload.position >= start
+				&& (end === null
+					|| payload.position <= end)),
+		)
+		return filteredPayloadArray
+	}
+
+	/**
+	 * Convert the PayloadArray to a vanilla Array.
+	 *
+	 * @return {Array} The resulting Array.
+	 */
+	toArray = () => this.payloads
```

Comment:
Something about this feels weird to me. Is the intent that this returns a reference to `this.payloads`, or a copy of it? I think the name `toArray()` makes me think it should be a copy (in which case this should return `[...this.payloads]`), whereas `asArray()` sounds more like the original object but in the form of a native array. So like:

```javascript
const payloadArray = new PayloadArray(…)

// I would expect this NOT to modify the original payload array:
// (in fact it's effectively pointless; you'd really assign `payloadArray.toArray()` to a variable first)
payloadArray.toArray().sort()

// I would expect this TO modify the original array:
payloadArray.asArray().sort()
```

So essentially I think this method just needs its intent and function clarified, I think?

Reply:
The intent is return a copy of it, but realizing that might not be what is actually happening.

We don't want the internal `payloads` to ever be changed outside of this class (because then everything about the class might break, since it assumes order for instance).

I'm going to add a test that ensures a copy while I fix this!